README.md
[![Gem Version](https://badge.fury.io/rb/stomp.svg)](https://badge.fury.io/rb/stomp)
[![Build Status](https://travis-ci.org/bbatsov/rubocop.svg?branch=master)](https://travis-ci.org/stompgem/stomp)
[![Code Climate](https://codeclimate.com/github/stompgem/stomp/badges/gpa.svg)](https://codeclimate.com/github/stompgem/stomp)

# README

* [Project Information](https://github.com/stompgem/stomp)

## Overview

An implementation of the Stomp protocol for Ruby. See:

* [STOMP 1.0](http://stomp.github.io/stomp-specification-1.0.html)
* [STOMP 1.1](http://stomp.github.io/stomp-specification-1.1.html)
* [STOMP 1.2](http://stomp.github.io/stomp-specification-1.2.html)

## Hash Login Example Usage (**this is the recommended login technique**):

```ruby
hash = {
  hosts: [
    # First connect is to remotehost1
    { login: 'login1', passcode: 'passcode1', host: 'remotehost1', port: 61_612, ssl: true },
    # First failover connect is to remotehost2
    { login: 'login2', passcode: 'passcode2', host: 'remotehost2', port: 61_613, ssl: false }
  ],
  # These are the default parameters and do not need to be set
  reliable: true,                  # reliable (use failover)
  initial_reconnect_delay: 0.01,   # initial delay before reconnect (secs)
  max_reconnect_delay: 30.0,       # max delay before reconnect
  use_exponential_back_off: true,  # increase delay between reconnect attpempts
  back_off_multiplier: 2,          # next delay multiplier
  max_reconnect_attempts: 0,       # retry forever, use # for maximum attempts
  randomize: false,                # do not radomize hosts hash before reconnect
  connect_timeout: 0,              # Timeout for TCP/TLS connects, use # for max seconds
  connect_headers: {},             # user supplied CONNECT headers (req'd for Stomp 1.1+)
  parse_timeout: 5,                # IO::select wait time on socket reads
  logger: nil,                     # user suplied callback logger instance
  dmh: false,                      # do not support multihomed IPV4 / IPV6 hosts during failover
  closed_check: true,              # check first if closed in each protocol method
  hbser: false,                    # raise on heartbeat send exception
  stompconn: false,                # Use STOMP instead of CONNECT
  usecrlf: false,                  # Use CRLF command and header line ends (1.2+)
  max_hbread_fails: 0,             # Max HB read fails before retry.  0 => never retry
  max_hbrlck_fails: 0,             # Max HB read lock obtain fails before retry.  0 => never retry
  fast_hbs_adjust: 0.0,            # Fast heartbeat senders sleep adjustment, seconds, needed ...
  # For fast heartbeat senders.  'fast' == YMMV.  If not
  # correct for your environment, expect unnecessary fail overs
  connread_timeout: 0,             # Timeout during CONNECT for read of CONNECTED/ERROR, secs
  tcp_nodelay: true,               # Turns on the TCP_NODELAY socket option; disables Nagle's algorithm
  start_timeout: 0,                # Timeout around Stomp::Client initialization
  sslctx_newparm: nil,             # Param for SSLContext.new
  ssl_post_conn_check: true,       # Further verify broker identity
  nto_cmd_read: true,              # No timeout on COMMAND read
}

# for a client
client = Stomp::Client.new(hash)

# for a connection
connection = Stomp::Connection.new(hash)
```

### Positional Parameter Usage:

```ruby
client = Stomp::Client.new("user", "pass", "localhost", 61613)
client.publish("/queue/mine", "hello world!")
client.subscribe("/queue/mine") do |msg|
    p msg
end
```

### Stomp URL Usage:

A Stomp URL must begin with `stomp://` and can be in one of the following forms:

```
stomp://host:port
stomp://host.domain.tld:port
stomp://login:passcode@host:port
stomp://login:passcode@host.domain.tld:port

# e.g. c = Stomp::Client.new(urlstring)
```

### Failover + SSL Example URL Usage:

```ruby
options = 'initialReconnectDelay=5000&randomize=false&useExponentialBackOff=false'
# remotehost1 uses SSL, remotehost2 doesn't
client = Stomp::Client.new("failover:(stomp+ssl://login1:passcode1@remotehost1:61612,stomp://login2:passcode2@remotehost2:61613)?#{options}")
client.publish('/queue/mine', 'hello world!')
client.subscribe('/queue/mine') do |msg|
  p msg
end
```

### New:

See _CHANGELOG.rdoc_ for details.

* Gem version 1.4.9. Fix two issues, enhance debugging and examples.
* Gem version 1.4.8. Fix missed merge in 1.4.7 release.
* Gem version 1.4.7. Add support for text SSL certs.  Do not use, use 1.4.8 instead.
* Gem version 1.4.6. Fix version 1.4.5 which breaks JRuby support.
* Gem version 1.4.5. JRuby broken here.  Use is not recommended.
* Gem version 1.4.4. Miscellaneous fixes, see CHANGELOG.md for details.
* Gem version 1.4.3. Fix broken install.  Do not try to install 1.4.2.
* Gem version 1.4.2. Fix memory leak, and others !: see CHANGELOG.md for details.
* Gem version 1.4.1. Important SSL changes !: see CHANGELOG.md for details.
* Gem version 1.4.0. Note: Change sementics of :parse_timeout, see CHANGELOG.md for details.
* Gem version 1.3.5. Miscellaneous fixes, see CHANGELOG.md for details.

For changes in older versions see CHANGELOG.rdoc for details.

### Historical Information:

Up until March 2009 the project was maintained and primarily developed by Brian McCallister.

### Source Code and Project URLs:

  [Source Code and Project](https://github.com/stompgem/stomp)

### Stomp Protocol Information:

  [Protocol Information](http://stomp.github.com/index.html)

#### Contributors

See CONTRIBUTORS.md.



CHANGELOG.md
# Stomp Gem Change Log

## 1.4.10 20200819

* Fix issue #165
* Reinstate autoflush behavior
* Correct _receive method signature
* Update version

* More debugging messages in netio.rb.

## 1.4.9 20191116

* More debugging messages in netio.rb.
* Test workaround for Artemis-2289.
* New put and get examples.
* Fix issue #160.
* Address issue #155.
* Suppress SSL stacktrace to stdout.
* Example logger enhancements.
* Optional STOMP frame in examples.

## 1.4.8 20181219

* Fix missed merge from 1.4.7 release.
* Add global debug output flag to all unit tests.

## 1.4.7 20181210

* Support SSL cert/key in text format.
* Test adjustments for Artemis
* Correct missing variable name
* Fix Ruby 2.6 test indentation warning

## 1.4.6 20181105

* Fix broken JRuby support.  Issue #153.
* Rework many tests for JRuby support.
* Add optional debug output to all tests.

## 1.4.5 20181027

* Add option for no timeout on network read
* Lock check of @io.ready for SSL sockets
* Squelch some deprecation warnings
* Address issue #146
* Reintroduce named excepton on nil header
* Environmentally control some debug flags
* Raise named exception on bad server frame
* Attempt to address issue #152
* Eliminate Ruby 2.5 warnings in unit tests
* Major refactor of all gem examples

## 1.4.4 20170611

* Merge fix for issue 141.
* Fix broken MatReconnectAttempts
* Attempt to detect mismatched SSL port and params.
* Issue 139, add ssl checks to receive timeout logic.
* Eliminate assert_nothing_raised from tests.
* Add Artemis suport to standard test bed.
* Run JRuby tests more easily.

## 1.4.3 20160821

* Quick fix of install failures.  Do not try to install 1.4.2.

## 1.4.2 20160820

* Refine SSL examples.
* Address issue #124 with additional RDOC.
* spec for Stomp::Client - check that headers passed to connection contain 
  required values as well as given custom and that given hash is not modified.
* Stomp::Client now does not modify given headers hash
* spec description enhancement.
* fix build_subscription_id - symbol and string were mixed up.
* STOMP_TESTSSL flag should enable all SSL tests.
* Add a basic Gemfile.
* Fix a memory leak in receipt implementation.
* Add unit test helper script.

## 1.4.1 20160623

* Add call to #post_connection_check to normal SSL processing.  This change
  further validates the name of the broker connected to.  This change adds to
  the current SSL connection processing logic, and is **highly recommended**.  In the
  case a client cannot tolerate this logic, it can be disabled by adding
  :ssl_post_conn_check => false to the connection hash.
* Fix typo in SSL failure recovery processing.

## 1.4.0 20160608

* Connection parameter :parse_timeout now means IO:select wait time for socket
    reads.  Consumer clients should see a significantly reduced memory
    footprint.  If the default (5 seconds) is not used by your client,
    we suggest you test carefully.
* Add example programs for sending / receiving large messages.
* Changelog format is changed from .rdoc to .md.
* README format is changed from .rdoc to .md.
* README format change of contributor's list.
* Add example utilities for generating the contributor's list.
* Eliminate many warnings when using 'rake test', mostly from the 2.x Ruby series.
* Get rakefile up to date.
* Add the 'adhoc' subdirectory, an area for experiments and issue recreation code.

## 1.3.5 20160302

* Add AMQ specific durable topic example.
* Output error to stderr only in logger is undefined.
* Move README changelog lower.
* Handle newline at start of receive buffer.
* Use Timeout::timeout instead of deprecated kernel version.
* If socket open on reconnect, close it before new open.
* On misc_err, make error messages more readable.
* Attempt to support both Rspec 2.14.1+ and 3.x.

## 1.3.4 20141202

* Change :start_timeout default to 0 (might break some clients) (#98).
* Allow user set of SSLContext options (#105).
* Allow user set of parm in SSLContext.new(parm) (#105).

## 1.3.3 20140810

* Do not attempt to write empty message bodies.
* Explicity close ssl socket on connection timeout.
* Fix incorrect behavior for empty header keys (#93)
* Do not override explicit :reliable => false.
* Fix client fail-over fails (#98)

## 1.3.2 20131208

* Anon tests assigned unique class name.
* Fix TypeError on connect timeout with 1.8.x, 2.x.
* Complete revert to previous logger logic.
* start_timeout and tcp_nodelay parameters
* SSL Fix, revert not setting default ciphers.
* Copy hash params at init.
* Fix ssl => true for Ruby 1.9.x and 2.x.
* Expanded list of STOMP default SSL ciphers:
* Do not change caller's :hosts array
* Issue #78, again.
* Clean up logger interfacing.
* Fixes from RSpec testing

## 1.3.1 20131002

* Method calls to the logger object should check for that method first (#83)

## 1.3.0 20130930

* ERROR frames now raise an exception in the Stomp::Client thread(#73, #81)
* Allow anonymous connections (#75)
* Fix for subscription id handling in STOMP 1.1 (#78)
* Added a NullLogger (#77)
* Added :tcp_nodelay option (disable Nagle's algorithm) (#76)
* Read receipt ids are now UUIDs
* Added a :start_timeout parameter

## 1.2.16 20130812

* Stomp::Client's should expose connection's host params

## 1.2.15 20130809

* Add user-specified timeout for initial CONNECTED/ERROR frame read.
* Eliminate dup Timeout::timeout in ssl connect
* Add license information to gemspec (#69)

## 1.2.14 20130819

* Version bump (1.2.13 release had Stomp::Version of 1.1.12.)
* Prevent dup subscription header on re-receive

## 1.2.13 20130817

* Issue #68, Stomp::Client#unreceive max_redeliveries off-by-one error

## 1.2.12 20130811

* Fix infinite loop when max reconn attempts is reached
* Enhance JRuby support in tests
* Issue #63, nil message on rapid AMQ restarts
* Issue #63, fast spurious failovers with JRuby and AMQ
* Issue #67, SSL SNI support (thanks Hiram)
* Proper cleanup when not reliable adn EOF from broker
* Remove extraneous privte declarations

lib/stomp/client.rb
# -*- encoding: utf-8 -*-

require 'thread'
require 'digest/sha1'
require 'timeout'
require 'forwardable'

module Stomp

  # Typical Stomp client class. Uses a listener thread to receive frames
  # from the server, any thread can send.
  #
  # Receives all happen in one thread, so consider not doing much processing
  # in that thread if you have much message volume.
  class Client
    extend Forwardable

    # Parameters hash
    attr_reader :parameters

    def_delegators :@connection, :login, :passcode, :port, :host, :ssl
    def_delegator :@parameters, :reliable

    # A new Client object can be initialized using three forms:
    #
    # Hash (this is the recommended Client initialization method):
    #
    #   hash = {
    #     :hosts => [
    #       {:login => "login1", :passcode => "passcode1", :host => "localhost", :port => 61616, :ssl => false},
    #       {:login => "login2", :passcode => "passcode2", :host => "remotehost", :port => 61617, :ssl => false}
    #     ],
    #     # These are the default parameters and do not need to be set
    #     :reliable => true,                  # reliable (use failover)
    #     :initial_reconnect_delay => 0.01,   # initial delay before reconnect (secs)
    #     :max_reconnect_delay => 30.0,       # max delay before reconnect
    #     :use_exponential_back_off => true,  # increase delay between reconnect attpempts
    #     :back_off_multiplier => 2,          # next delay multiplier
    #     :max_reconnect_attempts => 0,       # retry forever, use # for maximum attempts
    #     :randomize => false,                # do not radomize hosts hash before reconnect
    #     :connect_timeout => 0,              # Timeout for TCP/TLS connects, use # for max seconds
    #     :connect_headers => {},             # user supplied CONNECT headers (req'd for Stomp 1.1+)
    #     :parse_timeout => 5,                # IO::select wait time on socket reads
    #     :logger => nil,                     # user suplied callback logger instance
    #     :dmh => false,                      # do not support multihomed IPV4 / IPV6 hosts during failover
    #     :closed_check => true,              # check first if closed in each protocol method
    #     :hbser => false,                    # raise on heartbeat send exception
    #     :stompconn => false,                # Use STOMP instead of CONNECT
    #     :usecrlf => false,                  # Use CRLF command and header line ends (1.2+)
    #     :max_hbread_fails => 0,             # Max HB read fails before retry.  0 => never retry
    #     :max_hbrlck_fails => 0,             # Max HB read lock obtain fails before retry.  0 => never retry
    #     :fast_hbs_adjust => 0.0,            # Fast heartbeat senders sleep adjustment, seconds, needed ...
    #                                         # For fast heartbeat senders.  'fast' == YMMV.  If not
    #                                         # correct for your environment, expect unnecessary fail overs
    #     :connread_timeout => 0,             # Timeout during CONNECT for read of CONNECTED/ERROR, secs
    #     :tcp_nodelay => true,               # Turns on the TCP_NODELAY socket option; disables Nagle's algorithm
    #     :start_timeout => 0,                # Timeout around Stomp::Client initialization
    #     :sslctx_newparm => nil,             # Param for SSLContext.new
    #     :ssl_post_conn_check => true,       # Further verify broker identity
    #     :nto_cmd_read => true,              # No timeout on COMMAND read
    #   }
    #
    #   e.g. c = Stomp::Client.new(hash)
    #
    # Positional parameters:
    #   login     (String,  default : '')
    #   passcode  (String,  default : '')
    #   host      (String,  default : 'localhost')
    #   port      (Integer, default : 61613)
    #   reliable  (Boolean, default : false)
    #
    #   e.g. c = Stomp::Client.new('login', 'passcode', 'localhost', 61613, true)
    #
    # Stomp URL :
    #   A Stomp URL must begin with 'stomp://' and can be in one of the following forms:
    #
    #   stomp://host:port
    #   stomp://host.domain.tld:port
    #   stomp://login:passcode@host:port
    #   stomp://login:passcode@host.domain.tld:port
    #
    #   e.g. c = Stomp::Client.new(urlstring)
    #
    def initialize(login = '', passcode = '', host = 'localhost', port = 61613, reliable = false, autoflush = false)
      parse_hash_params(login) ||
        parse_stomp_url(login) ||
        parse_failover_url(login) ||
        parse_positional_params(login, passcode, host, port, reliable)

      @logger = @parameters[:logger] ||= Stomp::NullLogger.new
      @start_timeout = @parameters[:start_timeout] || 0
      @parameters[:client_main] = Thread::current
      ## p [ "CLINDBG", @parameters[:client_main] ]
      check_arguments!()

      # p [ "cldbg01", @parameters ]

      begin
        Timeout::timeout(@start_timeout) {
          create_error_handler
          create_connection(autoflush)
          start_listeners()
        }
      rescue Timeout::Error
        # p [ "cldbg02" ]
        ex = Stomp::Error::StartTimeoutException.new(@start_timeout)
        raise ex
      end
    end

    def create_error_handler
      client_thread = Thread.current
      if client_thread.respond_to?(:report_on_exception=)
        client_thread.report_on_exception=false
      end

      @error_listener = lambda do |error|
        exception = case error.body
                      when /ResourceAllocationException/i
                        Stomp::Error::ProducerFlowControlException.new(error)
                      when /ProtocolException/i
                        Stomp::Error::ProtocolException.new(error)
                      else
                        Stomp::Error::BrokerException.new(error)
                    end

        @receipt_listeners.delete(error.headers['receipt-id']) if error.headers['receipt-id']
        client_thread.raise exception
      end
    end

    def create_connection(autoflush)
      # p [ "ccon01", @parameters ]
      @connection = Connection.new(@parameters)
      @connection.autoflush = autoflush
    end
    private :create_connection

    # open is syntactic sugar for 'Client.new', see 'initialize' for usage.
    def self.open(login = '', passcode = '', host = 'localhost', port = 61613, reliable = false)
      Client.new(login, passcode, host, port, reliable)
    end

    # join the listener thread for this client,
    # generally used to wait for a quit signal.
    def join(limit = nil)
      @listener_thread.join(limit)
    end

    # Begin starts work in a a transaction by name.
    def begin(name, headers = {})
      @connection.begin(name, headers)
    end

    # Abort aborts work in a transaction by name.
    def abort(name, headers = {})
      @connection.abort(name, headers)

      # replay any ack'd messages in this transaction
      replay_list = @replay_messages_by_txn[name]
      if replay_list
        replay_list.each do |message|
          find_listener(message) # find_listener also calls the listener
        end
      end
    end

    # Commit commits work in a transaction by name.
    def commit(name, headers = {})
      txn_id = headers[:transaction]
      @replay_messages_by_txn.delete(txn_id)
      @connection.commit(name, headers)
    end

    # Subscribe to a destination, must be passed a block
    # which will be used as a callback listener.
    # Accepts a transaction header ( :transaction => 'some_transaction_id' ).
    def subscribe(destination, headers = {})
      raise Stomp::Error::NoListenerGiven unless block_given?
      headers = headers.symbolize_keys
      raise Stomp::Error::DestinationRequired unless destination
      # use subscription id to correlate messages to subscription. As described in
      # the SUBSCRIPTION section of the protocol: http://stomp.github.com/.
      # If no subscription id is provided, generate one.
      headers = headers.merge(:id => build_subscription_id(destination, headers))
      if @listeners[headers[:id]]
        raise Stomp::Error::DuplicateSubscription
      end
      @listeners[headers[:id]] = lambda {|msg| yield msg}
      @connection.subscribe(destination, headers)
    end

    # Unsubscribe from a subscription by name.
    def unsubscribe(destination, headers = {})
      headers = headers.symbolize_keys
      raise Stomp::Error::DestinationRequired unless destination
      headers = headers.merge(:id => build_subscription_id(destination, headers))
      @connection.unsubscribe(destination, headers)
      @listeners[headers[:id]] = nil
    end

    # Acknowledge a message, used when a subscription has specified
    # client acknowledgement ( connection.subscribe("/queue/a",{:ack => 'client'}).
    # Accepts a transaction header ( :transaction => 'some_transaction_id' ).
    def ack(message, headers = {})
      txn_id = headers[:transaction]
      if txn_id
        # lets keep around messages ack'd in this transaction in case we rollback
        replay_list = @replay_messages_by_txn[txn_id]
        if replay_list.nil?
          replay_list = []
          @replay_messages_by_txn[txn_id] = replay_list
        end
        replay_list << message
      end
      if block_given?
        headers = headers.merge(:receipt => register_receipt_listener(lambda {|r| yield r}))
      end
      context = ack_context_for(message, headers)
      @connection.ack context[:message_id], context[:headers]
    end

    # For posterity, we alias:
    alias acknowledge ack

    # Stomp 1.1+ NACK.
    def nack(message, headers = {})
      context = ack_context_for(message, headers)
      @connection.nack context[:message_id], context[:headers]
    end

    #
    def ack_context_for(message, headers)
      id = case protocol
        when Stomp::SPL_12
         'ack'
        when Stomp::SPL_11
         headers = headers.merge(:subscription => message.headers['subscription'])
         'message-id'
        else
         'message-id'
      end
      {:message_id => message.headers[id], :headers => headers}
    end

    # Unreceive a message, sending it back to its queue or to the DLQ.
    def unreceive(message, options = {})
      @connection.unreceive(message, options)
    end

    # Publishes message to destination.
    # If a block is given a receipt will be requested and passed to the
    # block on receipt.
    # Accepts a transaction header ( :transaction => 'some_transaction_id' ).
    def publish(destination, message, headers = {})
      headers = headers.symbolize_keys
      raise Stomp::Error::DestinationRequired unless destination
      if block_given?
        headers = headers.merge(:receipt => register_receipt_listener(lambda {|r| yield r}))
      end
      @connection.publish(destination, message, headers)
    end

    # Return the broker's CONNECTED frame to the client.  Misnamed.
    def connection_frame()
      @connection.connection_frame
    end

    # Return any RECEIPT frame received by DISCONNECT.
    def disconnect_receipt()
      @connection.disconnect_receipt
    end

    # open? tests if this client connection is open.
    def open?
      @connection.open?()
    end

    # close? tests if this client connection is closed.
    def closed?()
      @connection.closed?()
    end

    # jruby? tests if the connection has detcted a JRuby environment
    def jruby?()
      @connection.jruby
    end

    # close frees resources in use by this client.  The listener thread is
    # terminated, and disconnect on the connection is called.
    def close(headers={})
      @listener_thread.exit
      @connection.disconnect(headers)
    end

    # running checks if the thread was created and is not dead.
    def running()
      @listener_thread && !!@listener_thread.status
    end

    # set_logger identifies a new callback logger.
    def set_logger(logger)
      @logger = logger
      @connection.set_logger(logger)
    end

    # protocol returns the current client's protocol level.
    def protocol()
      @connection.protocol()
    end

    # valid_utf8? validates any given string for UTF8 compliance.
    def valid_utf8?(s)
      @connection.valid_utf8?(s)
    end

    # sha1 returns a SHA1 sum of a given string.
    def sha1(data)
      @connection.sha1(data)
    end

    # uuid returns a type 4 UUID.
    def uuid()
      @connection.uuid()
    end

    # hbsend_interval returns the connection's heartbeat send interval.
    def hbsend_interval()
      @connection.hbsend_interval()
    end

    # hbrecv_interval returns the connection's heartbeat receive interval.
    def hbrecv_interval()
      @connection.hbrecv_interval()
    end

    # hbsend_count returns the current connection's heartbeat send count.
    def hbsend_count()
      @connection.hbsend_count()
    end

    # hbrecv_count returns the current connection's heartbeat receive count.
    def hbrecv_count()
      @connection.hbrecv_count()
    end

    # Poll for asynchronous messages issued by broker.
    # Return nil of no message available, else the message
    def poll()
      @connection.poll()
    end

    # autoflush= sets the current connection's autoflush setting.
    def autoflush=(af)
      @connection.autoflush = af
    end

    # autoflush returns the current connection's autoflush setting.
    def autoflush()
      @connection.autoflush()
    end

  end # Class

end # Module
